### ISSUE TITLE
(SSR) Layout changes after client hydration

### ISSUE THREAD

**Issue by bchilcott** [bvaughn/react-resizable-panels#144]
**Edit:** I got this completely wrong. This problem isn't related to storage, it just so happens that `storage` changes the behaviour. The issue is that panel sizes aren't computed until hydration, which causes a layout shift. This issue does have some interesting discussion about SSR and persistence though.

Using the `storage` prop with the new NextJS app directory in a client component causes this error on the SSR step as localStorage is not defined - which is to be expected:
```
-- error src\components\Panels.tsx (46:48) @ localStorage
-- error ReferenceError: localStorage is not defined
```
The panel group is set up like this:
```js
<PanelGroup direction="horizontal" storage={localStorage}>
```
The lack of localStorage on the server doesn't seem to actually cause a problem, at least not an obvious one. Only the error which I'd rather not have.

**Comment by bchilcott** [comment]
Reproduced here
https://github.com/bchilcott/react-resizable-panels-144-repro

**Comment by bvaughn** [comment]
Thank you for this example! Very helpful.

There are two separate things going on here.

## The server error

This first one is easy to fix. Your example code is passing `localStorage` as a prop. You can fix the server side error by accounting for it being undefined:
```js
  let storage;
  try {
    storage = localStorage;
  } catch (error) {}

  return (
    <PanelGroup direction="horizontal" storage={storage}>
```

Or better yet, just don't pass anything– since `PanelGroup` defaults to use `localStorage` (and won't try to access it server side). All you really need to do is specify an `autoSaveId` prop 😄 

## The flickering bug

That being said, there is a flicker after server rendering (between the default size and the async local storage size). Replay [5c0489d5-2845-4413-8b7d-69542d50516c](https://app.replay.io/recording/react-resizable-panels-144--5c0489d5-2845-4413-8b7d-69542d50516c) shows this bug happening.

Here is a Loom I recorded where I looked into the issue and kind of just thought out loud:
https://www.loom.com/share/42d87330d06a483896266bc3b831cdd4

tl;dr is that I think you'd need some kind of custom persistence for this– that saved values to both `localStorage` and remotely (if you want to avoid the size flicker for the initial render. (This is psuedo code)
```js
function Example() {
  const [sizeA, sizeB] = getDefaultSizes();

  return (
    <PanelGroup autoSaveId="whatever" storage={customStorage}>
      <Panel defaultSize={sizeA} />
      <PanelResizeHandle />
      <Panel defaultSize={sizeB} />
    </PanelGroup>
  )
}

function getDefaultSizes() {
  if (typeof window === undefined) {
    // Read from DB
  } else {
    // Read from localStorage (or fall back to undefined)
  }
}

const customStorage = {
  getItem: (key: string) => {
    localStorage.getItem(key);
  },
  setItem: (key: string, value: string) => {
    localStorage.setItem(key, value);

    // Do some fetch() API call
  }
}
```

**Comment by bchilcott** [comment]
Wow - you have no idea how valuable this. Thanks so much for the explanation.

I was definitely using storage wrong - I wasn't too worried about actually persisting the sizes, just making the flickering go away. Out of interest, why does specifying `storage={localStorage}` with no `autoSaveId` stop the flickering from happening?

**Comment by bvaughn** [comment]
I'm glad it's helpful!

> why does specifying storage={localStorage} with no autoSaveId stop the flickering from happening?

`PanelGroup` will only persist the size if an `autoSaveId` prop is specified 😄

**Comment by bchilcott** [comment]
Sorry, I'm probably misunderstanding it. Does that mean there should be no functional difference between the following, if no `autoSaveId` is specified?
```jsx
<PanelGroup direction="horizontal">
```
```jsx
<PanelGroup direction="horizontal" storage={localStorage}>
```
Because using the first example I get this layout shifting from 50/50 to the specified default sizes:
![no-storage](https://github.com/bvaughn/react-resizable-panels/assets/38051283/529616e6-ac4c-4e58-9712-c4e90ca56056)
And with the second example, although the sizes obviously aren't persisted, they initialize to their defaults immediately:
![image](https://github.com/bvaughn/react-resizable-panels/assets/38051283/dec20225-d90e-42f9-9fa2-6c9521c65976)
I wouldn't expect specifying a `storage` option to make any difference here.

**Comment by bvaughn** [comment]
We are talking about two "defaults"

One default comes from the size you pass to `<Panel defaultSize>` and gets used when the user first sees the panel group.

If you pass an `autoSaveId` to `PanelGroup`– the second "default" is when a user resizes a panel, and then returns to the page later. In this case, the group should show the previous sizes the user chose.

That's the flicking with SSR, for reasons I mentioned in the Loom. The reason you don't see this if you just pass `storage` is because– without an `autoSaveId`– the `PanelGroup` doesn't try to save or restore the sizes.

**Comment by bchilcott** [comment]
Ah, got you. Thanks. Enjoy your coffees 😉

**Comment by bvaughn** [comment]
Oh! Thank you :)

**Comment by bchilcott** [comment]
Okay, either I'm just totally misunderstanding, or I'm having a different issue altogether. The fix you suggested fixes the localStorage undefined error, but brings the flickering back:
```jsx
// Flickering happens
  let storage;
  try {
    storage = localStorage;
  } catch (error) {}

  return (
    <PanelGroup direction="horizontal" storage={storage}>
```
```jsx
// Flickering happens
  return (
    <PanelGroup direction="horizontal">
```
```jsx
// Flickering happens
  return (
    <PanelGroup direction="horizontal" autoSaveId="something">
```
```jsx
// Flickering *doesn't* happen, but localStorage undefined error
  return (
    <PanelGroup direction="horizontal" storage={localStorage}>
```
To be clear, I'm not worried about persistence. As long as the values provided to `Panel` in `defaultSize` are the first thing the user sees, that's fine. I don't *actually* need the sizes to be persisted in localStorage or anywhere, it just so happens that passing `storage={localStorage}` to `PanelGroup` stops the layout from shifting. I understand that if I *did* want persistence, the sizes would have to be stored somewhere accessible to the server.

All I really want is to do is the following, with no flicker on the initial render:
```jsx
<PanelGroup direction="horizontal">
```

**Comment by bvaughn** [comment]
> Sorry this has been a bit of a rollercoaster of me confusing myself. I feel like `storage` shouldn't even be playing a part here.

Agreed! In hindsight, I should have spotted this last night but it was late 😄 

What's actually happening is that the `PanelGroup` doesn't compute its default sizes until it mounts:

https://github.com/bvaughn/react-resizable-panels/blob/a5108d9db370711dedc01a043718f1ee384223c1/packages/react-resizable-panels/src/PanelGroup.ts#L230-L233

I think I need to find a way to move this initialization logic into the render.

**Comment by bchilcott** [comment]
> but it was late 😄 

Yeah.. 4am clearly isn't the best time for me to do this 😅 Good to know though, and the discussion around SSR and storage is still good to have. 

I'll think about this some more.

**Comment by bvaughn** [comment]
I think I have an idea that could work. Will try to test it out soon.

**Comment by bvaughn** [comment]
Maybe it wouldn’t need to be required _except_ for SSR use cases. I think it will work though.

I’m traveling today but I’ll try to get an update published for this within the next day or two.

**Comment by bvaughn** [comment]
I think 259d794 should fix this, provided you specify a `defaultSize` prop for all `Panel`s.

**Comment by bvaughn** [comment]
And 264f4c5 will detect and warn about this case if you forget that prop:
![Screen Shot 2023-05-15 at 8 49 41 AM](https://github.com/bvaughn/react-resizable-panels/assets/29597/91de50e8-f4b5-454a-94e2-a968db9fcef5)

**Comment by bvaughn** [comment]
This fix will ship in 0.0.44 in a few minutes.

Thanks so much for the repro project. It was very helpful for me to verify this fix.

**Comment by bvaughn** [comment]
0.0.45 released

**Comment by bvaughn** [comment]
Btw I added an FAQ section here showing how to use a Cookie to avoid "flicker" for persisted layouts:
https://github.com/bvaughn/react-resizable-panels#how-can-i-use-persistent-layouts-with-ssr
